src/GShark/Geometry/Mesh.cs

```diff
@@ -548,6 +548,127 @@ private FaceData CountFaceEdges()
+
+            var CloserPointToFace_List = new List<Point3>();
+
+            foreach (MeshFace face in AllFaces)
+            {
+                // Retrives the Vertices associated with each face
+                List<MeshVertex> GSVertices = face.AdjacentVertices();
+                // Converts from MeshVertex to Point3
+                List<Point3> FacePoints = GSVertices.ConvertAll(v => (Point3)v);
+
+                if (GSVertices.Count == 3)
+                {
+                    var trianglePoints = new Point3[] { FacePoints[0], FacePoints[1], FacePoints[2] };
+                    Point3 ClosestPoint0 = ClosestPointToTriangle(trianglePoints, point);
+                    CloserPointToFace_List.Add(ClosestPoint0);
+                }
+                else if (GSVertices.Count == 4)
+                {
+                    var trianglePoints1 = new Point3[] { FacePoints[0], FacePoints[1], FacePoints[2] };
+                    var trianglePoints2 = new Point3[] { FacePoints[1], FacePoints[2], FacePoints[3] };
+                    var trianglePoints3 = new Point3[] { FacePoints[0], FacePoints[1], FacePoints[3] };
+                    var trianglePoints4 = new Point3[] { FacePoints[0], FacePoints[2], FacePoints[3] };
+
+                    var ClosestPoint1 = ClosestPointToTriangle(trianglePoints1, point);
+                    var ClosestPoint2 = ClosestPointToTriangle(trianglePoints2, point);
+                    var ClosestPoint3 = ClosestPointToTriangle(trianglePoints3, point);
+                    var ClosestPoint4 = ClosestPointToTriangle(trianglePoints4, point);
+
+                    var TrianglesCP = new Point3[4] { ClosestPoint1, ClosestPoint2, ClosestPoint3, ClosestPoint4 };
+                    Point3 ClosestPoint = point.CloudClosestPoint(TrianglesCP);
+
+                    CloserPointToFace_List.Add(ClosestPoint);
+                }
+                else 
+                {
+                    //throw new Exception("Ngon detected");
+                    // Stellate method (from Ngon to triangles) 
+                    List<Point3> GSVerticesPoints = GSVertices.ConvertAll(v => (Point3)v);
+                    Point3 ngonCentre = Point3.AveragePoint(GSVerticesPoints);
+
+                    var TrianglesCP = new List<Point3>();
+
+                    for (int i = 0; i < GSVertices.Count; i++)
+                    {
+                        Point3[] trianglePoints = new Point3[] { ngonCentre, GSVertices[i], GSVertices[(i + 1)% GSVertices.Count] };
+                        Point3 ClosestPointToNgon = ClosestPointToTriangle(trianglePoints, point);
+                        TrianglesCP.Add(ClosestPointToNgon);
+                    }
+                    Point3 ClosestPoint = point.CloudClosestPoint(TrianglesCP);
+                    CloserPointToFace_List.Add(ClosestPoint);
+                }
+            }
+            Point3 MeshClosestPoint = point.CloudClosestPoint(CloserPointToFace_List);
+            return MeshClosestPoint;
```

Comment:
Since `CloudClosestPoint` could return a MeshVertex (since we don't convert to points to speed up performance), then we don't want this function to return the same object point that define the mesh. If the person using this function is then changing the point that this function returns, he/she could potentially modify the mesh itself.
Just return a copy of the point found:

```suggestion
            Point3 meshClosestPoint = point.CloudClosestPoint(CloserPointToFace_List);
            return new Point3(meshClosestPoint);
```